src/benchmarklib/table_builder.hpp

```diff
@@ -0,0 +1,240 @@
+
+template <typename T, bool _has_value>
+class optional_constexpr<T, _has_value, std::enable_if_t<!_has_value>> {
+ public:
+  template <typename... Args>
+  explicit optional_constexpr(Args&&...) {}
+
+  static_assert(!_has_value);
+  static constexpr bool has_value = false;
+
+  T& value() {
+    Assert(false, "empty optional has no value");
+    return {};
+  }
+};
+
+template <typename T, typename Enable = void>
+struct is_optional : std::false_type {};
+
+template <typename T>
+struct is_optional<T, std::enable_if_t<std::is_same_v<T, std::optional<typename T::value_type>>>> : std::true_type {};
+
+// is_optional_v<T> <=> T is of type std::optional<?>
+template <typename T>
+constexpr bool is_optional_v = is_optional<T>::value;
+
+template <typename T, typename Enable = void>
+struct GetValueType {
+  using value_type = T;
+};
+
+template <typename T>
+struct GetValueType<T, std::enable_if_t<is_optional_v<T>>> {
+  using value_type = typename T::value_type;
+};
+
+// get the value_type (lets call it V) of std::optional<V> and V using the same syntax - needed for templating
+template <typename T>
+using get_value_type = typename GetValueType<T>::value_type;
+
+// check if std::optional<V> or V is null using the same syntax - needed for templating
+template <typename T>
+constexpr bool is_null(const T& maybe_optional) {
+  if constexpr (is_optional_v<T>) {
+    return !maybe_optional.has_value();
+  } else {
+    return false;
+  }
+}
+
+// get the value of std::optional<V> or V using the same syntax - needed for templating
+template <typename T>
+get_value_type<T>& get_value(T& maybe_optional) {
+  if constexpr (is_optional_v<T>) {
+    return maybe_optional.value();
+  } else {
+    return maybe_optional;
+  }
+}
+
+}  // namespace
+
+namespace opossum {
+
+/**
+ * Helper to build a table with a static column layout, specified by ctor args types and names. Keeps a
+ * data vector for each column and appends values to them in append_row(). For nullable columns an additional
+ * is_null_vector is kept. Automatically creates chunks in accordance with the specified chunk size.
+ */
+template <typename... DataTypes>
+class TableBuilder {
+ public:
+  // names is a list of strings defining column names, types is a list of equal length defining respective column types
+  // types may contain std::optional<?>, which will result in a nullable column, otherwise columns are not nullable
+  template <typename... Names>
+  TableBuilder(size_t chunk_size, const hana::tuple<DataTypes...>& types, const hana::tuple<Names...>& names,
+               opossum::UseMvcc use_mvcc, size_t estimated_rows = 0)
+      : _use_mvcc(use_mvcc), _estimated_rows_per_chunk(std::min(estimated_rows, chunk_size)) {
+    BOOST_HANA_CONSTANT_ASSERT(hana::size(names) == hana::size(types));
+
+    // Iterate over the column types/names and create the columns.
+    auto column_definitions = TableColumnDefinitions{};
+    hana::fold_left(hana::zip(names, types), column_definitions,
```

Comment:
That worked after also adding a ` -> std::decay_t<decltype(definitions)> ` to the lambda's return type, but how does that help with the awkward duality? Do you prefer calling the lmbda's parameter `column_definitions` too? That would (kind of) shadow `column_definitions` from outside of the lambda and probably confuse readers in my opinion.

Reply:
compiler/tidy probably won't even allow shadowing. sad that we cannot avoid it. I'd call it "column_definitions" and "column_definitions2" then, better not to hide these awkward things. But that's up to you.

Reply:
Yep, because of [-Werror,-Wshadow-uncaptured-local] the commit can't be built right now by jenkins (on the sidon server that was not an issue). I didn't even consider that shadowing, since an uncaptured local variable is not even in the lambda's scope. Oh well, I will go with "column_definitions2" for the lambda's parameter.

Reply:
I found another way: capturing column_definitions into the lambda and just using for_each. I guess we were looking for a complex solution, but there was a very simple one...